src/python/WMCore/MicroService/MSTransferor/RequestInfo.py

```diff
@@ -64,28 +80,55 @@ def __init__(self, msConfig, rucioObj, logger):
         self.rucioToken = None
         self.tokenValidity = None
         self.openRunning = self.msConfig["openRunning"]
+        self.pileupDocs = []
 
-    def __call__(self, reqRecords):
+    def __call__(self, reqRecords, pileupDocs):
         """
-        Run the unified transferor box
+        Execute the main logic for input data placement, which depends on:
+        1. parsing the workflow and finding the parent dataset name, if any
+        2. check the expected locations for pileup dataset(s)
+        3. discover all the input data blocks, which are valid
+        4. create a map of child to parent blocks
         :param reqRecords: input records
-        :return: output records
+        :param pileupDocs: list with dictionary of pileup objects
+        :return: output records. Note that in case of failures, it
+            will be a subset of the input records.
         """
+        self.pileupDocs = pileupDocs
         # obtain new unified Configuration
         uConfig = self.unifiedConfig()
         if not uConfig:
             self.logger.warning("Failed to fetch the latest unified config. Skipping this cycle")
             return []
         self.logger.info("Going to process %d requests.", len(reqRecords))
 
+        # setup the Rucio token
+        self.setupRucio()
+
         # create a Workflow object representing the request, matching
         # against some specific templates
         workflows = self.classifyWorkflows(reqRecords)
 
-        # setup the Rucio token
-        self.setupRucio()
-        # get complete requests information (based on Unified Transferor logic)
-        self.unified(workflows)
+        # Step 1: figure out any possible parent datasets
+        self.logger.info("Getting/setting parent datasets for %d requests", len(workflows))
+        with CodeTimer("### getParentDatasets", logger=self.logger):
```

Comment:
I"m not sure if it will be the case, but each code with CodeTimer assumes short execution while you only obtain rucio token once. Since token has validity interval and depending how long request will go it may be wise to obtain token before each `with CodeTimer` statement, or at least check token timestamp and have some default to avoid expired token. In ideal world token should be short lived and our calls to services may be long standing queries. It is just a suggestion though.

Reply:
We haven't spotted any issues with the current implementation, but it does not mean that the token will never expiry in between those operations.
That's a good suggestion and can be easily addressed, I am updating this code in a bit. Thanks!